### PR TITLE
TSQR: Automatically detect CUBLAS & CUSOLVER TPLs; improve TPL handle wrappers

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/tpetra @trilinos/belos @trilinos/anasazi 
@iyamazaki @vqd8a @ndellingwood 

## Motivation

1. Automatically detect CUBLAS and CUSOLVER TPLs.  Now users don't have to set `TPL_ENABLE_CUBLAS=ON` etc. explicitly.
2. Improve cuBLAS and cuSOLVER handle wrappers so that the kokkos-kernels team can adopt the design pattern directly.
3. Fix issue with noncontiguous R in `CuSolverNodeTsqr::extract_R` observed by @iyamazaki .

## Related Issues

* Supersedes PR #6496.
* Related to https://github.com/trilinos/Trilinos/pull/6488, https://github.com/kokkos/kokkos-kernels/pull/552, https://github.com/kokkos/kokkos-kernels/pull/554

## Testing

My CUDA build automatically detects the CUBLAS and CUSOLVER TPLs.

### CODE REVIEW

File: packages/tpetra/tsqr/cmake/Dependencies.cmake [hunk 1]
```diff
@@ -6,3 +6,10 @@ SET(LIB_REQUIRED_DEP_TPLS)
 SET(LIB_OPTIONAL_DEP_TPLS CUBLAS CUSOLVER)
 SET(TEST_REQUIRED_DEP_TPLS)
 SET(TEST_OPTIONAL_DEP_TPLS)
+
+IF(DEFINED TPL_ENABLE_CUDA)
```
Comment (bartlettroscoe) [inline packages/tpetra/tsqr/cmake/Dependencies.cmake]: You don't need this IF() statement, do you?  Don't you just need `IF (TPL_ENABLE_CUDA)`?  Is there some behavior of CMake you are seeing that I am not aware of?

Comment (mhoemmen) [reply packages/tpetra/tsqr/cmake/Dependencies.cmake]: @bartlettroscoe I'm just being excessively cautious, but if you like, I can remove the outer `IF`.

Comment (bartlettroscoe) [reply packages/tpetra/tsqr/cmake/Dependencies.cmake]: @mhoemmen, for better or for worse, CMake evaluates undefined variables as empty "" which evaluates to `FALSE` so:

```
IF (DEFINED <SOME_VAR>)
  IF (<SOME_VAR>)
     ...
  ENDIF()
ENDIF()
```

and

```
IF (<SOME_VAR>)
     ...
ENDIF()
```

should be identical

Comment (mhoemmen) [reply packages/tpetra/tsqr/cmake/Dependencies.cmake]: @bartlettroscoe This was a problem historically when we did things like this:
```
IF(FOO AND BAR)
```
where FOO was not defined but BAR was.  I'll change this, though.

**Comment by mhoemmen** [comment]
@trilinos/framework All the jobs errored out.  The CUDA 9.2 build said this at the end of its configuration step:
```
-- Configuring done
CMake Warning:
  Value of Trilinos_ENABLE_TESTS contained a newline; truncating


CMake Warning:
  Value of Trilinos_PARALLEL_LINK_JOBS_LIMIT contained a newline; truncating


-- Generating done
CMake Warning:
  Value of Trilinos_ENABLE_TESTS contained a newline; truncating


CMake Warning:
  Value of Trilinos_PARALLEL_LINK_JOBS_LIMIT contained a newline; truncating


-- Build files have been written to: /home/trilinos/jenkins/ride/workspace/trilinos-folder/Trilinos_pullrequest_cuda_9.2/pull_request_test
```

**Comment by mhoemmen** [comment]
@trilinos/tpetra Could someone please approve this?  It fixes a build warning that is hindering SPARC's warnings-as-errors build.

**Comment by bartlettroscoe** [comment]
@mhoemmen, I would approve this but I think an official Tpetra developer should review and approve if possible.  If none is able to do so, then I can approve.  Just let me know.

**Comment by mhoemmen** [comment]
@bartlettroscoe wrote:
>  I would approve this but I think an official Tpetra developer should review and approve if possible. If none is able to do so, then I can approve. Just let me know.

Could you please approve?  None of the other Tpetra developers know much about this code.  Thanks! :-D

**Pull request merged**